### DISCUSSION TITLE
Issues with nested table using purrr package--converted TS objects have datatype character instead of dbl

### DISCUSSION THREAD

**Discussion by princessblackswan** [teamalgoritma/community#806] [Q&A] [answered]
Hi all,

Hoping someone with experience with nested table to help me out.

I have nested table with 45 rows, and I would like to convert train data to time series objects *ts* and *msts* for fitting. So that has been done with no issue, however I notice that the params list contains character type S3:ts objects, which is out of the ordinary because the previous times I have done this the datatype is usually dbl. As a result I'm unable to fit the models, because of this non-numeric TS objects. I tried a workaround to use lapply for all elements inside the list to convert them to numeric, but when this is done its no longer a time series object.

Hoping someone can help!  Thank you

```
#Nest the new tables
walmart_improve_nested <- walmart_data_improve %>%
  group_by(Store) %>%
  nest()

#Cross Validation
walmart_improve_nested <- walmart_improve_nested %>%
  mutate(train = map(data, train_split),
         test = map(data, test_split))

#Decompose and plot
walmart_ts_decompose <- function(df){
  df$Weekly_Sales = as.numeric(df$Weekly_Sales)
  ts(df$Weekly_Sales, frequency=52) %>% decompose() %>% autoplot()
}

walmart_msts_decompose <- function(df){
  df$Weekly_Sales = as.numeric(df$Weekly_Sales)
  msts(df$Weekly_Sales, seasonal.periods = c(4, 12, 26, 52)) %>% mstl() %>% autoplot()
}

#Visualize decomposed plots 
walmart_improve_nested <- walmart_improve_nested %>% 
  mutate(ts_decompose = map(train, walmart_ts_decompose),
         msts_decompose = map(train, walmart_msts_decompose))

#Create list of time series object functions
ts_function <- list(
  ts_obj = function(x) ts(as.numeric(x$Weekly_Sales), frequency=52),
  msts_obj = function(x) msts(as.numeric(x$Weekly_Sales), seasonal.periods = c(4,12,26,52))
)


#As there are two objects, replicate each unique Store number twice i.e one for each time series object
ts_function<- ts_function %>% 
  rep(length(unique(walmart$Store))) %>% 
  enframe("ts_func_name", "ts_func") %>% 
  mutate(Store = sort(rep(unique(walmart$Store), length(unique(.$ts_func_name)))))


#left join 
walmart_improve_nested <- left_join(walmart_improve_nested, ts_function) 

model_function <- list(
   stlm = function(x) stlm(x),
   tbats = function(x) tbats(x),
   HoltWinters = function(x) HoltWinters(x)
   )

model_function<- model_function %>% 
  rep(length(unique(walmart$Store))) %>% 
  enframe("model_name", "model_func") %>% 
  mutate(Store = sort(rep(unique(walmart$Store), length(unique(.$model_name)))))

#left join 
walmart_improve_nested <- left_join(walmart_improve_nested, model_function) 

walmart_improve_nested <- walmart_improve_nested %>%
  mutate(params = map(train, ~list(x = .x)))

walmart_improve_nested <- walmart_improve_nested %>% 
  mutate(
    data = invoke_map(ts_func, params),
    params = map(data, ~list(x = .x))
  )

```

The above code runs fine and are as expected with the exception of the class of the params list, which is character instead of dbl like usual. I have explicitly coerce the target variable as numeric but somehow the resulting TS object becomes character. I believe this is the cause of the problem as the other times when the type of params is 'dbl' there has been no issues.

Some screenshots
<img width="823" alt="Screenshot 2023-11-24 at 00 31 19" src="https://github.com/teamalgoritma/community/assets/122248785/4f7ff8b0-ae44-4b2d-af37-233c4305c346">
<img width="878" alt="Screenshot 2023-11-24 at 00 33 45" src="https://github.com/teamalgoritma/community/assets/122248785/4606626f-8e64-4795-b579-684f8d5a8574">

Please note that the following code results in error:

`
walmart_improve_nested <- walmart_improve_nested %>% 
  mutate(
    fitted = invoke_map(model_func, params)
  )
`
Error message:

Error in `mutate()`:
ℹ In argument: `fitted = invoke_map(model_func, params)`.
ℹ In group 1: `Store = "1"`.
Caused by error in `map2()`:
ℹ In index: 1.
Caused by error in `+.default`:
! non-numeric argument to binary operator
Backtrace:
  1. walmart_improve_nested %>% ...
 17. global `<fn>`(x = `<ts>`)
 18. forecast::stlm(x)
 19. forecast::mstl(x, s.window = s.window, robust = robust)
 21. base::NextMethod(.Generic)

Thank you

**Comment by IrfanChairurrachman** [comment]
Dear @princessblackswan ,

Pertanyaan Anda sudah kami baca dan sedang kami telusuri permasalahannya. Harap menunggu untuk balasan selanjutnya, terima kasih banyak.

**Comment by hwulanayu** [comment, marked as answer]
It seems like the issue might be arising from the structure of the `params` list. Specifically, when you apply the `invoke_map` function to `model_func` and `params`, it might not be handling the character type of your time series objects correctly.

To resolve this, you can modify the code to ensure that the `params` list contains numeric time series objects. You can do this by directly applying your time series functions (`ts_obj` and `msts_obj`) to the data within the `train` list. Here's a modified version of your code:

```R
walmart_improve_nested <- walmart_improve_nested %>%
  mutate(params = map(train, ts_function$ts_obj))

# Continue with the rest of your code

walmart_improve_nested <- walmart_improve_nested %>%
  mutate(
    data = invoke_map(ts_func, params),
    params = map(data, ~list(x = .x))
  )
...
```

This modification directly applies `ts_function$ts_obj` to the `train` list, ensuring that the resulting `params` list contains numeric time series objects. After making this change, you should be able to proceed with the rest of your code without encountering issues related to non-numeric time series objects.

**Comment by kinantytasyao** [comment]
Halo @princessblackswan , karena tidak ada konfirmasi lebih lanjut dan untuk menjaga agar forum diskusi tetap berjalan kondusif, kami izin melakukan lock/close diskusi ini. Semoga dapat dipahami.
